File: 235-LowestCommonAncestorOfABinarySearchTree/235-LowestCommonAncestorOfABinarySearchTree.java
// Last updated: 8/16/2026, 9:51:02 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) return root;

        while (root != null) {
            if (p.val > root.val && q.val > root.val) {
                // go to right
                root = root.right;
            } else if (p.val < root.val && q.val < root.val) {
                // go to left
                root = root.left;
            } else {
                return root;
            }
        }
        return null;
    }
}

File: 235-LowestCommonAncestorofaBinarySearchTree/235-LowestCommonAncestorofaBinarySearchTree.py
# Last updated: 9/16/2026, 12:10:47 PM
1# Definition for a binary tree node.
2# class TreeNode:
3#     def __init__(self, x):
4#         self.val = x
5#         self.left = None
6#         self.right = None
7
8class Solution:
9    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
10        curr = root
11
12        while curr:
13            if curr.val > p.val and curr.val > q.val:
14                # go to left subtree
15                curr = curr.left
16            elif curr.val < p.val and curr.val < q.val:
17                # go to right
18                curr = curr.right
19            else:
20                return curr
21        
22        return None